File: packages/core-node/src/providers/wave.ts
import { createHmac, timingSafeEqual } from "node:crypto";

import {
  PaymentError,
  PaymentStatus,
  type PaymentEvent,
  type PaymentProvider,
  type PaymentRequest,
  type PaymentSession,
  type PaymentStatusResult,
  type RefundResult,
} from "../types.js";
import { validateRefundAmount } from "../refund-validation.js";
import { InMemoryWebhookEventStore, type WebhookEventStore } from "../webhook-event-store.js";

const WAVE_API_BASE_URL = "https://api.wave.com/v1";

export interface WaveProviderConfig {
  apiKey: string;
  webhookSecret: string;
  /** Override for local API-compatible mocks. Defaults to the Wave API. */
  baseUrl?: string;
  webhookEventStore?: WebhookEventStore;
}

interface WaveCheckoutSession {
  id?: string;
  amount?: string | number;
  currency?: string;
  client_reference?: string | null;
  wave_launch_url?: string;
  /**
   * Wave checkout lifecycle. It takes priority over payment_status for expiry
   * detection because payment_status has no documented "expired" value.
   */
  checkout_status?: "open" | "complete" | "expired";
  payment_status?: string;
  error_code?: string;
  when_expires?: string;
  when_completed?: string;
  when_created?: string;
}

interface WaveRefundResponse {
  id?: string;
  amount?: string | number;
  status?: string;
}

interface WaveErrorResponse {
  /** Wave may return a top-level `code` for API-level errors. */
  code?: string;
  error_code?: string;
  error_message?: string;
  message?: string;
}

interface WaveWebhookPayload {
  id?: string;
  type?: string;
  data?: WaveCheckoutSession;
}

/** Error intentionally reduced to the common PaymentError code. */
export class WaveProviderError extends Error {
  public constructor(public readonly code: PaymentError, message: string) {
    super(message);
    this.name = "WaveProviderError";
  }
}

export class WaveProvider implements PaymentProvider {
  private readonly webhookEventStore: WebhookEventStore;

  public constructor(private readonly config: WaveProviderConfig) {
    this.webhookEventStore = config.webhookEventStore ?? new InMemoryWebhookEventStore();
  }

  public async initiatePayment(params: PaymentRequest): Promise<PaymentSession> {
    const response = await this.request("/checkout/sessions", {
      method: "POST",
      body: JSON.stringify({
        amount: params.amount,
        currency: "XOF",
        ...(params.failureUrl === undefined ? {} : { error_url: params.failureUrl }),
        ...(params.successUrl === undefined ? {} : { success_url: params.successUrl }),
        client_reference: params.reference,
      }),
    });
    const session = await this.readJson<WaveCheckoutSession>(response);

    if (typeof session.id !== "string" || session.id.length === 0) {
      throw new WaveProviderError(PaymentError.Unknown, "Wave checkout response is missing id");
    }

    return {
      id: session.id,
      reference: params.reference,
      amount: params.amount,
      currency: params.currency,
      status: PaymentStatus.Pending,
      ...(session.wave_launch_url === undefined ? {} : { paymentUrl: session.wave_launch_url }),
    };
  }

  public async checkStatus(sessionId: string): Promise<PaymentStatusResult> {
    const session = await this.getCheckoutSession(sessionId);

    // checkout_status is evaluated first: only it can explicitly represent expiry.
    if (session.checkout_status === "expired") {
      return { status: PaymentStatus.Expired };
    }
    if (typeof session.payment_status !== "string") {
      throw new WaveProviderError(PaymentError.Unknown, "Wave checkout response is missing payment_status");
    }
    return this.toPaymentStatusResult(session.payment_status, session.error_code);
  }

  public async handleWebhook(
    rawBody: string | Buffer,
    headers: Record<string, string | string[] | undefined>
  ): Promise<PaymentEvent> {
    const signature = this.findHeader(headers, "x-wave-signature");
    const rawBytes = typeof rawBody === "string" ? Buffer.from(rawBody) : rawBody;

    if (signature === undefined || !this.isValidSignature(rawBytes, signature)) {
      throw new WaveProviderError(PaymentError.Unknown, "Invalid Wave webhook signature");
    }

    const payload = this.parseWebhookPayload(rawBytes.toString("utf8"));
    const data = payload.data;
    const sessionId = data?.id;
    if (sessionId === undefined || sessionId.length === 0) {
      throw new WaveProviderError(PaymentError.Unknown, "Wave webhook is missing checkout session id");
    }

    const status = this.toWebhookStatus(payload.type, data?.checkout_status, data?.payment_status);
    const ref = data?.client_reference === null || data?.client_reference === undefined ? {} : { reference: data.client_reference };
    const occurredAt = data?.when_completed ?? data?.when_expires ?? data?.when_created ?? new Date().toISOString();
    const id = payload.id ?? sessionId;

    const event: PaymentEvent = status === PaymentStatus.Failed
      ? { id, sessionId, status: PaymentStatus.Failed, ...ref, occurredAt, error: this.mapError(200, data?.error_code) }
      : { id, sessionId, status, ...ref, occurredAt };
    return this.webhookEventStore.record(event);
  }

  public async refund(sessionId: string, amount?: number): Promise<RefundResult> {
    if (amount !== undefined) {
      validateRefundAmount(
        amount,
        (code, message) => new WaveProviderError(code, message)
      );
    }

    const session = await this.getCheckoutSession(sessionId);
    const originalAmount = this.parseOriginalAmount(session.amount);

    if (amount !== undefined && amount > originalAmount) {
      throw new WaveProviderError(
        PaymentError.RefundAmountExceedsBalance,
        "Refund amount exceeds the original payment amount"
      );
    }

    const refundAmount = amount ?? originalAmount;

    const response = await this.request(`/checkout/sessions/${encodeURIComponent(sessionId)}/refund`, {
      method: "POST",
      ...(amount === undefined ? {} : { body: JSON.stringify({ amount }) }),
    });
    const refund = await this.readJson<WaveRefundResponse>(response);

    if (typeof refund.id !== "string" || typeof refund.amount !== "number") {
      throw new WaveProviderError(PaymentError.Unknown, "Wave refund response is incomplete");
    }

    return {
      sessionId,
      refundId: refund.id,
      amount: refund.amount,
      status: refund.status === undefined ? PaymentStatus.Success : this.toPaymentStatus(refund.status),
    };
  }

  private async request(path: string, init: RequestInit = {}): Promise<Response> {
    let response: Response;
    try {
      response = await fetch(`${this.baseUrl()}${path}`, {
        ...init,
        headers: {
          Authorization: `Bearer ${this.config.apiKey}`,
          "Content-Type": "application/json",
          ...init.headers,
        },
      });
    } catch (error: unknown) {
      throw this.toProviderError(error);
    }

    if (!response.ok) {
      throw await this.toResponseError(response);
    }
    return response;
  }

  private baseUrl(): string {
    return (this.config.baseUrl ?? WAVE_API_BASE_URL).replace(/\/$/, "");
  }

  private async getCheckoutSession(sessionId: string): Promise<WaveCheckoutSession> {
    const response = await this.request(`/checkout/sessions/${encodeURIComponent(sessionId)}`);
    return this.readJson<WaveCheckoutSession>(response);
  }

  private parseOriginalAmount(amount: string | number | undefined): number {
    const parsed = typeof amount === "number" ? amount : Number(amount);
    if (!Number.isSafeInteger(parsed) || parsed <= 0) {
      throw new WaveProviderError(PaymentError.Unknown, "Wave checkout response is missing a valid amount");
    }
    return parsed;
  }

  private async toResponseError(response: Response): Promise<WaveProviderError> {
    const body = await this.readJsonOrEmpty<WaveErrorResponse>(response);
    const message = body.error_message ?? body.message ?? `Wave request failed with HTTP ${response.status}`;
    return new WaveProviderError(this.mapError(response.status, body.error_code ?? body.code), message);
  }

  private toProviderError(error: unknown): WaveProviderError {
    if (error instanceof WaveProviderError) {
      return error;
    }
    return new WaveProviderError(
      PaymentError.ProviderTimeout,
      error instanceof Error ? error.message : "Wave request failed"
    );
  }

  private mapError(httpStatus: number, waveCode: string | undefined): PaymentError {
    if (waveCode === "insufficient-funds") {
      return PaymentError.InsufficientFunds;
    }
    if (waveCode === "payer-mobile-mismatch" || waveCode === "invalid-phone") {
      return PaymentError.InvalidPhone;
    }
    if (waveCode === "payment-cancelled" || waveCode === "user-cancelled") {
      return PaymentError.UserCancelled;
    }
    if (httpStatus === 408 || httpStatus === 429 || httpStatus >= 500) {
      return PaymentError.ProviderTimeout;
    }
    return PaymentError.Unknown;
  }

  private toPaymentStatus(status: string): PaymentStatus {
    switch (status.toLowerCase()) {
      case "succeeded":
      case "success":
        return PaymentStatus.Success;
      case "processing":
      case "pending":
        return PaymentStatus.Pending;
      case "cancelled":
      case "failed":
        return PaymentStatus.Failed;
      default:
        throw new WaveProviderError(PaymentError.Unknown, "Unknown Wave payment status");
    }
  }

  private toPaymentStatusResult(status: string, errorCode?: string): PaymentStatusResult {
    const paymentStatus = this.toPaymentStatus(status);
    if (paymentStatus !== PaymentStatus.Failed) {
      return { status: paymentStatus };
    }

    return { status: paymentStatus, error: this.mapError(200, errorCode) };
  }

  private toWebhookStatus(
    type: string | undefined,
    checkoutStatus: WaveCheckoutSession["checkout_status"],
    paymentStatus: string | undefined
  ): PaymentStatus {
    // Wave currently documents no dedicated webhook event type for expiration.
    // Accept an expired checkout_status even when the event type is otherwise unknown.
    if (checkoutStatus === "expired") {
      return PaymentStatus.Expired;
    }
    if (type === "checkout.session.completed") {
      return PaymentStatus.Success;
    }
    if (type === "checkout.session.payment_failed") {
      return PaymentStatus.Failed;
    }
    if (paymentStatus === undefined) {
      throw new WaveProviderError(PaymentError.Unknown, "Wave webhook has an unknown event type");
    }
    return this.toPaymentStatus(paymentStatus);
  }

  private findHeader(
    headers: Record<string, string | string[] | undefined>,
    expectedName: string
  ): string | undefined {
    for (const [name, value] of Object.entries(headers)) {
      if (name.toLowerCase() !== expectedName) {
        continue;
      }
      if (typeof value === "string") {
        return value;
      }
      if (Array.isArray(value) && value.length === 1) {
        return value[0];
      }
    }
    return undefined;
  }

  private isValidSignature(rawBody: Buffer, headerValue: string): boolean {
    const expectedSignature = createHmac("sha256", this.config.webhookSecret).update(rawBody).digest("hex");
    const candidates = headerValue.split(",").map((part) => part.trim().replace(/^v1=/, ""));

    return candidates.some((candidate) => {
      const expected = Buffer.from(expectedSignature);
      const received = Buffer.from(candidate);
      return expected.length === received.length && timingSafeEqual(expected, received);
    });
  }

  private parseWebhookPayload(serializedBody: string): WaveWebhookPayload {
    try {
      const payload: unknown = JSON.parse(serializedBody);
      if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
        throw new Error("Webhook payload must be an object");
      }
      return payload as WaveWebhookPayload;
    } catch {
      throw new WaveProviderError(PaymentError.Unknown, "Invalid Wave webhook payload");
    }
  }

  private async readJson<T>(response: Response): Promise<T> {
    const body: unknown = await response.json();
    return body as T;
  }

  private async readJsonOrEmpty<T>(response: Response): Promise<T> {
    try {
      return await this.readJson<T>(response);
    } catch {
      return {} as T;
    }
  }
}

File: packages/core-python/waslpay/providers/wave.py
from __future__ import annotations

import hashlib
import hmac
import json
import sys
from typing import Any, Literal, TypedDict, cast

import httpx

from ..contracts import PaymentProvider, WebhookHeaders
from ..enums import PaymentError, PaymentStatus
from ..errors import ProviderError
from ..models import PaymentEvent, PaymentRequest, PaymentSession, PaymentStatusResult, RefundResult
from ..refund_validation import validate_refund_amount
from ..webhook_event_store import InMemoryWebhookEventStore, WebhookEventStore


class WaveCheckoutSession(TypedDict, total=False):
    id: str
    amount: str | int
    client_reference: str | None
    wave_launch_url: str
    checkout_status: Literal["open", "complete", "expired"]
    payment_status: str
    error_code: str
    when_expires: str
    when_completed: str
    when_created: str


class WaveProvider(PaymentProvider):
    _BASE_URL = "https://api.wave.com/v1"

    def __init__(self, client: httpx.AsyncClient, api_key: str, webhook_secret: str, webhook_event_store: WebhookEventStore | None = None, base_url: str | None = None) -> None:
        self._client, self._api_key, self._webhook_secret = client, api_key, webhook_secret
        self._base_url = (base_url if base_url is not None else self._BASE_URL).rstrip("/")
        # The in-memory default is instance-local; inject durable shared storage in production.
        self._webhook_event_store = (
            webhook_event_store
            if webhook_event_store is not None
            else InMemoryWebhookEventStore()
        )

    async def initiate_payment(self, params: PaymentRequest) -> PaymentSession:
        payload = await self._request("POST", "/checkout/sessions", json={"amount": params.amount, "currency": "XOF", "error_url": params.failure_url, "success_url": params.success_url, "client_reference": params.reference})
        session_id = payload.get("id")
        if not isinstance(session_id, str): raise ProviderError(PaymentError.UNKNOWN, "Wave checkout response is missing id")
        return PaymentSession(id=session_id, reference=params.reference, amount=params.amount, currency=params.currency, status=PaymentStatus.PENDING, payment_url=payload.get("wave_launch_url") if isinstance(payload.get("wave_launch_url"), str) else None)

    async def check_status(self, session_id: str) -> PaymentStatusResult:
        session = await self._checkout_session(session_id)
        # checkout_status takes priority because payment_status has no documented expired value.
        if session.get("checkout_status") == "expired":
            return PaymentStatusResult(status=PaymentStatus.EXPIRED)
        return self._status_result(session.get("payment_status"), session.get("error_code"))

    async def handle_webhook(self, raw_body: str | bytes, headers: WebhookHeaders) -> PaymentEvent:
        raw = raw_body.encode() if isinstance(raw_body, str) else raw_body
        signature = self._header(headers, "x-wave-signature")
        expected = hmac.new(self._webhook_secret.encode(), raw, hashlib.sha256).hexdigest()
        candidates = [part.strip().removeprefix("v1=") for part in signature.split(",")] if signature else []
        if not any(hmac.compare_digest(expected, candidate) for candidate in candidates): raise ProviderError(PaymentError.UNKNOWN, "Invalid Wave webhook signature")
        payload = self._decode(raw); data = payload.get("data")
        if not isinstance(data, dict) or not isinstance(data.get("id"), str): raise ProviderError(PaymentError.UNKNOWN, "Incomplete Wave webhook payload")
        event_type = payload.get("type")
        # Wave currently documents no dedicated webhook event type for expiration.
        # An expired checkout_status is therefore accepted before event-type mapping.
        status = (
            PaymentStatus.EXPIRED
            if data.get("checkout_status") == "expired"
            else PaymentStatus.SUCCESS if event_type == "checkout.session.completed"
            else PaymentStatus.FAILED if event_type == "checkout.session.payment_failed"
            else self._status(data.get("payment_status"))
        )
        error = self._map_error(200, data.get("error_code")) if status is PaymentStatus.FAILED else None
        event = PaymentEvent(id=str(payload.get("id", data["id"])), session_id=data["id"], status=status, occurred_at=str(data.get("when_completed", data.get("when_expires", data.get("when_created", "1970-01-01T00:00:00Z")))), reference=data.get("client_reference") if isinstance(data.get("client_reference"), str) else None, error=error)
        return self._webhook_event_store.process(event, self._process_webhook_event)

    @staticmethod
    def _process_webhook_event(event: PaymentEvent) -> PaymentEvent:
        return event

    async def refund(self, session_id: str, amount: int | float | None = None) -> RefundResult:
        if amount is not None:
            amount = validate_refund_amount(
                amount,
                lambda code, message: ProviderError(code, message),
            )

        session = await self._checkout_session(session_id)
        original_amount = self._original_amount(session.get("amount"))

        if amount is not None and amount > original_amount:
            raise ProviderError(
                PaymentError.REFUND_AMOUNT_EXCEEDS_BALANCE,
                "Refund amount exceeds the original payment amount",
            )

        refund_amount = amount if amount is not None else original_amount
        payload = await self._request("POST", f"/checkout/sessions/{session_id}/refund", json={} if amount is None else {"amount": refund_amount})
        refund_id, refund_amount = payload.get("id"), payload.get("amount")
        if not isinstance(refund_id, str) or not isinstance(refund_amount, int): raise ProviderError(PaymentError.UNKNOWN, "Incomplete Wave refund response")
        return RefundResult(session_id=session_id, refund_id=refund_id, amount=refund_amount, status=self._status(payload.get("status", "succeeded")))

    async def _checkout_session(self, session_id: str) -> WaveCheckoutSession:
        return cast(WaveCheckoutSession, await self._request("GET", f"/checkout/sessions/{session_id}"))

    async def _request(self, method: str, path: str, **kwargs: Any) -> dict[str, Any]:
        try: response = await self._client.request(method, self._base_url + path, headers={"Authorization": f"Bearer {self._api_key}"}, **kwargs)
        except httpx.HTTPError as exc: raise ProviderError(PaymentError.PROVIDER_TIMEOUT, str(exc)) from exc
        payload = response.json() if response.content else {}; payload = payload if isinstance(payload, dict) else {}
        if response.is_error:
            error = self._map_error(response.status_code, payload.get("error_code", payload.get("code")))
            raise ProviderError(error, str(payload.get("error_message", payload.get("message", "Wave request failed"))))
        return payload

    @staticmethod
    def _status(status: object) -> PaymentStatus:
        payment_status = {
            "succeeded": PaymentStatus.SUCCESS,
            "success": PaymentStatus.SUCCESS,
            "processing": PaymentStatus.PENDING,
            "pending": PaymentStatus.PENDING,
            "cancelled": PaymentStatus.FAILED,
            "failed": PaymentStatus.FAILED,
        }.get(str(status).lower())
        if payment_status is None:
            raise ProviderError(PaymentError.UNKNOWN, "Unknown Wave payment status")
        return payment_status

    @staticmethod
    def _original_amount(value: object) -> int:
        if isinstance(value, bool):
            raise ProviderError(PaymentError.UNKNOWN, "Wave checkout response is missing a valid amount")
        if isinstance(value, int) and 0 < value <= sys.maxsize:
            return value
        if isinstance(value, str) and value.isdigit():
            parsed = int(value)
            if 0 < parsed <= sys.maxsize:
                return parsed
        raise ProviderError(PaymentError.UNKNOWN, "Wave checkout response is missing a valid amount")

    def _status_result(self, status: object, error_code: object) -> PaymentStatusResult:
        payment_status = self._status(status)
        return PaymentStatusResult(status=payment_status, error=self._map_error(200, error_code) if payment_status is PaymentStatus.FAILED else None)

    @staticmethod
    def _map_error(http_status: int, code: object) -> PaymentError:
        normalized_code = str(code) if code is not None else ""
        if normalized_code == "insufficient-funds": return PaymentError.INSUFFICIENT_FUNDS
        if normalized_code in {"payer-mobile-mismatch", "invalid-phone"}: return PaymentError.INVALID_PHONE
        if normalized_code in {"payment-cancelled", "user-cancelled"}: return PaymentError.USER_CANCELLED
        if http_status >= 500 or http_status in {408, 429}: return PaymentError.PROVIDER_TIMEOUT
        return PaymentError.UNKNOWN

    @staticmethod
    def _decode(raw: bytes) -> dict[str, Any]:
        payload = json.loads(raw)
        if not isinstance(payload, dict): raise ProviderError(PaymentError.UNKNOWN, "Invalid Wave webhook payload")
        return payload

    @staticmethod
    def _header(headers: WebhookHeaders, name: str) -> str | None:
        return next((value for key, value in headers.items() if key.lower() == name and isinstance(value, str)), None)
